Wrappers/Python/ccpi/optimisation/functions/Function.py

```diff
@@ -23,55 +23,493 @@
+        
+        self.L = L
+        self.domain = domain
+
+    def __call__(self,x):
+        
+        r"""Returns the value of the function F at x: :math:`F(x)`
+        
+        """        
+        
         raise NotImplementedError
 
     def gradient(self, x, out=None):
-        '''Returns the gradient of the function at x, if the function is differentiable'''
+        
+        r"""Returns the value of the gradient of function F at x, if it is differentiable
+        
+        .. math:: F'(x)
+        
+        """
         raise NotImplementedError
 
     def proximal(self, x, tau, out=None):
-        '''This returns the proximal operator for the function at x, tau'''
+        
+        r"""Returns the proximal operator of function :math:`\tau F` at x        
+        .. math:: \mathrm{prox}_{\tau F}(x) = \underset{z}{\mathrm{argmin}} \frac{1}{2}\|z - x\|^{2} + \tau F(z)
+        """
         raise NotImplementedError
 
-    def convex_conjugate(self, x, out=None):
-        '''This evaluates the convex conjugate of the function at x'''
+    def convex_conjugate(self, x):
+        r""" Returns the convex conjugate of function :math:`F` at :math:`x^{*}`,
+        
+        .. math:: F^{*}(x^{*}) = \underset{x^{*}}{\sup} <x^{*}, x> - F(x)
+                
+        """
         raise NotImplementedError
 
     def proximal_conjugate(self, x, tau, out = None):
-        '''This returns the proximal operator for the convex conjugate of the function at x, tau'''
-        raise NotImplementedError
+        
+        r"""Returns the proximal operator of the convex conjugate of function :math:`\tau F` at :math:`x^{*}`
+        
+        .. math:: \mathrm{prox}_{\tau F^{*}}(x^{*}) = \underset{z^{*}}{\mathrm{argmin}} \frac{1}{2}\|z^{*} - x^{*}\|^{2} + \tau F^{*}(z^{*})
+        
+        Due to Moreau’s identity, we have an analytic formula to compute the proximal operator of the convex conjugate :math:`F^{*}`
+        
+        .. math:: \mathrm{prox}_{\tau F^{*}}(x) = x - \tau\mathrm{prox}_{\tau^{-1} F}(\tau^{-1}x)
+                
+        """
+        if out is None:
+            return x - tau * self.proximal(x/tau, 1/tau)
+        else:            
+            self.proximal(x/tau, 1/tau, out = out)
+            out*=-tau
+            out.add(x, out = out)                       
 
     def grad(self, x):
-        '''Alias of gradient(x,None)'''
+        """Alias of gradient(x,None)"""
         warnings.warn('''This method will disappear in following 
         versions of the CIL. Use gradient instead''', DeprecationWarning)
         return self.gradient(x, out=None)
 
     def prox(self, x, tau):
-        '''Alias of proximal(x, tau, None)'''
+        """ Alias of proximal(x, tau, None)"""
         warnings.warn('''This method will disappear in following 
         versions of the CIL. Use proximal instead''', DeprecationWarning)
         return self.proximal(x, tau, out=None)
+    
+    def domain(self):
```

Comment:
2 comments:

1. the member `domain` can be accessed. Why do you create a method to access it?
2. the member `domain` isn't used and we still haven't decided how to use it. Why add it in the first place?

Reply:
1. you are right
2. I would like to have it because it makes life a lot easier if we have. For all the functions that I defined is ok atm, but in the future we definitely we want to use it. This is the reason I commented out [this](https://github.com/vais-ral/CCPi-Framework/blob/new_Functions_version/Wrappers/Python/ccpi/optimisation/functions/L2NormSquared.py#L54)